### PR TITLE
move dataservice, dispatchers,  selectors into ngrx-data

### LINKED ISSUE #41: Consider how to clean up the entity and app data services (less developer code)
Ideas:
- put app-data entity service in ngrx data and have ngrx-data inject it, and if you dont like it you can replace it on your own

### LINKED ISSUE #52: Consolidate the dispatchers to a unified dispatcher
Move to ngrx-data

But document how to extend this for your own use

### PR DESCRIPTION
**Pull request by wardbell**
close #41. close #52

### CODE REVIEW

File: src/client/app/store/app-entity-store.module.ts [hunk 1]
```diff
@@ -1,42 +1,48 @@
...
+import { appConfigReducers, appConfigServices } from './app-config';
 
 // This has to be an object and not a new() due to AOT
-const initialEntityCollectionState: EntityCollection<any> = {
-  filter: '',
-  entities: [],
-  filteredEntities: [],
-  loading: false
-};
+const initialEntityCollectionState = new EntityCollection();
```
Comment (johnpapa) [inline src/client/app/store/app-entity-store.module.ts]: Our `TODO` reminds us that this cannot be a function as AOT won't complete successfully.

File: src/client/app/store/services/app.dispatchers.ts [hunk 2]
```diff
@@ -1,12 +0,0 @@
-import { Injectable } from '@angular/core';
```
Comment (johnpapa) [inline src/client/app/store/services/app.dispatchers.ts]: Great that we are removing this and the other files like this (dispatchers and selectors). Let's make sure this is added to the documentation so folks know how to extend dispatchers and selectors when desired.

File: src/client/app/store/app-entity-store.module.ts (hunk 1 above)
Comment (wardbell) [reply src/client/app/store/app-entity-store.module.ts]: Should be fine. It's not used in metadata. I came up with different way.

File: src/client/ngrx-data/basic-data.service.ts [hunk 3]
```diff
@@ -24,28 +28,28 @@ export interface BasicDataServiceOptions {
...
   protected entitiesUrl: string;
-  protected saveDelay: number;
-  protected getDelay: number;
+  protected getDelay: LetOp;
+  protected saveDelay: LetOp;
 
   constructor(
     protected http: HttpClient,
-    { api = '/api', entitiesName, entityName, getDelay = 0, saveDelay = 0 }: BasicDataServiceOptions
+    { api, entitiesName, entityName, getDelay = 0, saveDelay = 0 }: BasicDataServiceOptions
```
Comment (johnpapa) [inline src/client/ngrx-data/basic-data.service.ts]: api is here

**Comment by johnpapa** [comment]
This PR is taking the folder/file structure from this ...

image

to this

image

File: src/client/ngrx-data/entity-data.service.ts [hunk 4]
```diff
@@ -34,16 +49,14 @@ export class AppDataService implements EntityDataService {
...
-      : entityClass.name
-    ).toLowerCase();
+    const entityName = getEntityName(entityClass);
 
     let service = this.services[entityName];
 
     if (!service) {
       const entitiesName = this.pluralizer.pluralize(entityName);
       service = new BasicDataService(this.http, {
+        api: this.api,
```
Comment (johnpapa) [inline src/client/ngrx-data/entity-data.service.ts]: LGTM

**Comment by johnpapa** [comment]
revised image

image

**Pull request merged**